mmda/parsers/pdfplumber_parser.py

```diff
@@ -99,171 +100,200 @@ def __init__(
         self.split_at_punctuation = split_at_punctuation
 
     def parse(self, input_pdf_path: str) -> Document:
-        doc = self._load_pdf_as_doc(input_pdf_path)
-        return doc
-
-    def _load_page_tokens(
-        self,
-        page: pdfplumber.page.Page,
-        page_index: int,
-        x_tolerance: int = 1.5,
-        y_tolerance: int = 2,
-        keep_blank_chars: bool = False,
-        use_text_flow: bool = True,
-        horizontal_ltr: bool = True,
-        vertical_ttb: bool = True,
-        extra_attrs: Optional[List[str]] = None,
-    ):
-        token_data = page.extract_words(
-            x_tolerance=x_tolerance,
-            y_tolerance=y_tolerance,
-            keep_blank_chars=keep_blank_chars,
-            use_text_flow=use_text_flow,
-            horizontal_ltr=horizontal_ltr,
-            vertical_ttb=vertical_ttb,
-            extra_attrs=extra_attrs,
-            split_at_punctuation=self.split_at_punctuation
-        )
-        page_tokens = [
-            {
-                "text": token["text"],
-                "bbox": Box.from_pdf_coordinates(
-                    x1=float(token["x0"]),
-                    y1=float(token["top"]),
-                    x2=float(token["x1"]),
-                    y2=float(token["bottom"]),
-                    page_width=float(page.width), 
-                    page_height=float(page.height),
-                    page=int(page_index)
-                ).get_relative(
-                    page_width=float(page.width), page_height=float(page.height)
-                ),
-            }
-            for token in token_data
-        ]
-
-        return page_tokens
-
-    def _load_pdf_tokens(self, input_pdf_path: str) -> Dict:
         plumber_pdf_object = pdfplumber.open(input_pdf_path)
-        page_to_line_to_tokens = {}
-        for page_id in range(len(plumber_pdf_object.pages)):
-            cur_page = plumber_pdf_object.pages[page_id]
-            page_tokens = self._load_page_tokens(
-                page=cur_page,
-                page_index=page_id,
+        all_tokens = []
+        all_word_ids = []
+        last_word_id = -1
+        all_row_ids = []
+        last_row_id = -1
+        all_page_ids = []
+        for page_id, page in enumerate(plumber_pdf_object.pages):
+            # 1) tokens we use for Document.symbols
+            coarse_tokens = page.extract_words(
+                x_tolerance=self.token_x_tolerance,
+                y_tolerance=self.token_y_tolerance,
+                keep_blank_chars=self.keep_blank_chars,
+                use_text_flow=self.use_text_flow,
+                horizontal_ltr=self.horizontal_ltr,
+                vertical_ttb=self.vertical_ttb,
+                extra_attrs=self.extra_attrs,
+                split_at_punctuation=None
+            )
```

Comment:
is it "ok" resource-wise that we're gonna do `.extract_words` twice? Is this the part of PDFPlumber that takes a long time?

Reply:
i'm not sure if it's OK or not. but this is an acceptable fix for now. we can always profile & try a separate implementation in future if it seems like this is too inefficient, but my impression now is that the biggest gains in efficiency arent here but in the predictors